synapse/handlers/e2e_keys.py

```diff
@@ -977,12 +985,66 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
         Raises:
             NotFoundError: if the key is not found
         """
+        user = UserID.from_string(user_id)
         key = yield self.store.get_e2e_cross_signing_key(
             user_id, key_type, from_user_id
         )
+
+        if key is None and self.is_mine(user):
+            # Attempt to fetch the missing key from the remote user's server
+            try:
+                remote_result = yield self.federation.query_client_keys(
```

Comment:
Might be better to use `query_user_devices` which would be easier to parse as it only queries one user at a time so you don't need to check the user ID in the return value.  It will return something like:
```json
{
  "user_id": "@alice:example.com",
  "stream_id": 5,
  "devices": [
     // ...
  ],
  "master_key": {
    "user_id": "@alice:example.com",
    "usage": ["master"],
    "keys": {
      "ed25519:base64+master+public+key": "base64+master+public+key"
    }
  },
  "self_signing_key": {
    "user_id": "@alice:example.com",
    "usage": ["self_signing"],
    "keys": {
      "ed25519:base64+self+signing+public+key": "base64+self+signing+public+key"
    },
    "signatures": {
      "@alice:example.com": {
        "ed25519:base64+master+public+key": "base64+signature"
      }
    }
  }
}
```

Reply:
Oh, that is indeed very useful.